src/main/java/org/folio/sidecar/integration/keycloak/IntrospectionService.java

```diff
@@ -0,0 +1,114 @@
+package org.folio.sidecar.integration.keycloak;
+
+import static io.vertx.core.Future.failedFuture;
+import static io.vertx.core.Future.succeededFuture;
+import static java.lang.String.join;
+import static org.folio.sidecar.integration.kafka.LogoutEvent.Type.LOGOUT_ALL;
+import static org.folio.sidecar.integration.keycloak.model.TokenIntrospectionResponse.INACTIVE_TOKEN;
+import static org.folio.sidecar.integration.okapi.OkapiHeaders.TOKEN;
+import static org.folio.sidecar.utils.JwtUtils.getSessionIdClaim;
+import static org.folio.sidecar.utils.JwtUtils.getUserIdClaim;
+import static org.folio.sidecar.utils.RoutingUtils.getOriginTenant;
+import static org.folio.sidecar.utils.RoutingUtils.getParsedToken;
+import static org.folio.sidecar.utils.SecureStoreUtils.tenantStoreKey;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import io.quarkus.security.UnauthorizedException;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.client.HttpResponse;
+import jakarta.enterprise.context.ApplicationScoped;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.sidecar.integration.kafka.LogoutEvent;
+import org.folio.sidecar.integration.keycloak.configuration.KeycloakProperties;
+import org.folio.sidecar.integration.keycloak.model.TokenIntrospectionResponse;
+import org.folio.sidecar.model.ClientCredentials;
+import org.folio.sidecar.service.CacheInvalidatable;
+import org.folio.sidecar.service.store.AsyncSecureStore;
+
+@Log4j2
+@ApplicationScoped
+@RequiredArgsConstructor
+public class IntrospectionService implements CacheInvalidatable {
+
+  private final KeycloakProperties properties;
+  private final KeycloakClient keycloakClient;
+  private final AsyncSecureStore secureStore;
+  private final Cache<String, TokenIntrospectionResponse> tokenCache;
+
+  public Future<RoutingContext> checkActiveToken(RoutingContext ctx) {
+    return introspectToken(ctx)
+      .map(TokenIntrospectionResponse::isActive)
+      .flatMap(IntrospectionService::handleTokenStatus);
+  }
+
+  @Override
+  public void invalidate(LogoutEvent event) {
+    if (LOGOUT_ALL == event.getType()) {
+      tokenCache.asMap().forEach((key, value) -> {
+        if (key.contains(event.getUserId())) {
+          tokenCache.put(key, INACTIVE_TOKEN);
+        }
+      });
+      return;
+    }
+
+    tokenCache.asMap().forEach((key, value) -> {
+      if (key.contains(event.getUserId()) && key.contains(event.getSessionId())) {
```

Comment:
Does this logic make sense?
Previously we already replaced all tokens where `key.contains(event.getUserId())` , here we do the same but with a more specific condition - but it's already done above